File: comparator.py
import copy 

items = [
    {
        'weight': [0.1, 0],
    },
    {
        'weight': [1, 0],
    }
]

def compare_stats(current_item: dict, items: list[dict]):
    _current_item = current_item
    organized_stats = organize_stats(items)
    for (key, (val, _)) in current_item.items():
        stats = organized_stats[key]
        average =  sum(stats) / len(stats)
        abs_diff = val - average

        _current_item[key][1] = round((abs_diff / average) * 100, 2)
    return _current_item

def organize_stats(items: list[dict]):
    results = {}
    for item in items:
        for (key, (val, _)) in item.items():
            if not results.get(key, False):
                results[key] = []
            results[key].append(val)

    return results

for (i, item) in enumerate(items):
    _items = copy.deepcopy(items)
    _items.pop(i)

    print(compare_stats(item, _items))

File: frontend/src/utils/comparator.ts
import type { Item } from "../stores/types"

const NEGATIVE_ATTRS = [
    "AimingTime",
    "RecoilDelay",
    "ReloadTime"
]

export function calculateDifferences(currentItem: Item, items: Item[]) {
    const differences: Record<string, number> = {};
    const [keys, organizedStats] = organizeStats(items); 
    keys.forEach(key => {
        const stats = organizedStats[key];
        const average = stats.reduce((sum, a) => sum + a) / stats.length;
        const abs_diff = (currentItem[key] as number) - average;

        let diff = (abs_diff / average) * 100;
        if(!Number.isNaN(diff)) {
            if(!Number.isFinite(diff))
                diff = 100;
            if(NEGATIVE_ATTRS.includes(key))
                diff = diff * -1;
            differences[key] = Math.round(diff);
        }
    });

    return differences;
}

export function organizeStats(items: Item[]): [string[], Record<string, number[]>] {
    const results: Record<string, number[]> = {};
    const keys: string[] = [];

    items.forEach((item) => {
        Object.keys(item).forEach(key => {
            const val = item[key];
            if(typeof(val) === 'number') {
                if(!results[key]) {
                    results[key] = [];
                };

                results[key].push(val);
                keys.push(key);
            };
        });
    });

    return [keys, results];
}
